pkg/operator/ceph/file/mds.go

```diff
@@ -14,194 +14,189 @@ See the License for the specific language governing permissions and
 	"github.com/rook/rook/pkg/clusterd"
 	"github.com/rook/rook/pkg/daemon/ceph/client"
 	mdsdaemon "github.com/rook/rook/pkg/daemon/ceph/mds"
-	"github.com/rook/rook/pkg/daemon/ceph/model"
-	opmon "github.com/rook/rook/pkg/operator/ceph/cluster/mon"
-	"github.com/rook/rook/pkg/operator/ceph/pool"
 	"github.com/rook/rook/pkg/operator/k8sutil"
 	"k8s.io/api/core/v1"
-	extensions "k8s.io/api/extensions/v1beta1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
 )
 
 const (
+	// AppName is the name of Rook's Ceph mds (File) sub-app
 	AppName = "rook-ceph-mds"
-)
 
-// Create the file system
-func CreateFilesystem(context *clusterd.Context, fs cephv1beta1.Filesystem, version string, hostNetwork bool, ownerRefs []metav1.OwnerReference) error {
-	if err := validateFilesystem(context, fs); err != nil {
-		return err
-	}
+	keyringSecretKeyName = "keyring"
 
-	var dataPools []*model.Pool
-	for _, p := range fs.Spec.DataPools {
-		dataPools = append(dataPools, p.ToModel(""))
-	}
-	f := mdsdaemon.NewFS(fs.Name, fs.Spec.MetadataPool.ToModel(""), dataPools, fs.Spec.MetadataServer.ActiveCount)
-	if err := f.CreateFilesystem(context, fs.Namespace); err != nil {
-		return fmt.Errorf("failed to create file system %s: %+v", fs.Name, err)
-	}
+	// timeout if mds is not ready for upgrade after some time
+	fsUpgradePrepareTimeout = 3 * time.Minute
+)
 
-	filesystem, err := client.GetFilesystem(context, fs.Namespace, fs.Name)
-	if err != nil {
-		return fmt.Errorf("failed to get file system %s. %+v", fs.Name, err)
+type cluster struct {
+	context     *clusterd.Context
+	Version     string
+	HostNetwork bool
+	fs          cephv1beta1.Filesystem
+	fsID        string
+	ownerRefs   []metav1.OwnerReference
+}
+
+func newCluster(
+	context *clusterd.Context,
+	version string,
+	hostNetwork bool,
+	fs cephv1beta1.Filesystem,
+	fsdetails *client.CephFilesystemDetails,
+	ownerRefs []metav1.OwnerReference,
+) *cluster {
+	return &cluster{
+		context:     context,
+		Version:     version,
+		HostNetwork: hostNetwork,
+		fs:          fs,
+		fsID:        strconv.Itoa(fsdetails.ID),
+		ownerRefs:   ownerRefs,
 	}
+}
 
-	logger.Infof("start running mds for file system %s", fs.Name)
+type mdsConfig struct {
+	ResourceName string
+	DaemonName   string
+}
 
-	// start the deployment
-	deployment := makeDeployment(context.Clientset, fs, strconv.Itoa(filesystem.ID), version, hostNetwork, ownerRefs)
-	_, err = context.Clientset.ExtensionsV1beta1().Deployments(fs.Namespace).Create(deployment)
-	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create mds deployment. %+v", err)
-		}
-		logger.Infof("mds deployment %s already exists", deployment.Name)
-	} else {
-		logger.Infof("mds deployment %s started", deployment.Name)
+// return true if an attempt was made to prepare the filesystem associated for upgrade
+func (c *cluster) deleteLegacyMdsDeployment() bool {
```

Comment:
You made a comment on first pass about whether I could just attempt deletion of the legacy deployment without checking on it, and while I'm still sure we need to check on it, I realized from my work last night that the legacy deployment's name is known to us, and we don't actually have to loop through all the matching `rook_file_system=<fsname>` deployments to find the legacy one. That simplified this function a lot, and I'm getting set up for another upgrade test here in my dev cluster to verify.